src/storage/compression/bitpacking.cpp

```diff
@@ -0,0 +1,599 @@
+
+	void LoadDecompressFunction() {
+		decompress_function = &BitpackingPrimitives::UnPackBlock<PRE_CAST_TYPE>;
+		return;
+	}
+
+	idx_t position_in_group = 0;
+	data_ptr_t current_width_group_ptr;
+	data_ptr_t bitpacking_width_ptr;
+	bitpacking_width_t current_width;
+
+	PhysicalType type;
+
+	PRE_CAST_TYPE decompression_buffer[BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE];
+};
+
+template <class T, class PRE_CAST_TYPE>
+unique_ptr<SegmentScanState> BitpackingInitScan(ColumnSegment &segment) {
+	auto result = make_unique<BitpackingScanState<T, PRE_CAST_TYPE>>(segment);
+	return move(result);
+}
+
+//===--------------------------------------------------------------------===//
+// Scan base data
+//===--------------------------------------------------------------------===//
+template <class T, class PRE_CAST_TYPE>
+void BitpackingScanPartial(ColumnSegment &segment, ColumnScanState &state, idx_t scan_count, Vector &result,
+                           idx_t result_offset) {
+	auto &scan_state = (BitpackingScanState<T, PRE_CAST_TYPE> &)*state.scan_state;
+
+	T *result_data = FlatVector::GetData<T>(result);
+	result.SetVectorType(VectorType::FLAT_VECTOR);
+
+	idx_t scanned = 0;
+
+	while (scanned < scan_count) {
+		// Exhausted this width group, move pointers to next group and load bitwidth for next group.
+		if (scan_state.position_in_group >= BITPACKING_WIDTH_GROUP_SIZE) {
+			scan_state.position_in_group = 0;
+			scan_state.bitpacking_width_ptr -= sizeof(bitpacking_width_t);
+			scan_state.current_width_group_ptr += (scan_state.current_width * BITPACKING_WIDTH_GROUP_SIZE) / 8;
+			scan_state.LoadCurrentBitWidth();
+		}
+
+		idx_t offset_in_compression_group =
+		    scan_state.position_in_group % BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE;
+
+		idx_t to_scan = MinValue<idx_t>(scan_count - scanned, BitpackingPrimitives::BITPACKING_ALGORITHM_GROUP_SIZE -
+		                                                          offset_in_compression_group);
+
+		// TODO Optimization: avoid use of decompression buffer if we can compress straight to result vector
+		// Calculate start of compression algorithm group
+		data_ptr_t current_position_ptr =
+		    scan_state.current_width_group_ptr + scan_state.position_in_group * scan_state.current_width / 8;
+		data_ptr_t decompression_group_start_pointer =
+		    current_position_ptr - offset_in_compression_group * scan_state.current_width / 8;
+
+		// Decompress compression algorithm to buffer
+		scan_state.decompress_function((data_ptr_t)scan_state.decompression_buffer, decompression_group_start_pointer,
+		                               scan_state.current_width);
+
+		// Copy decompressed result to vector
+		T *current_result_ptr = result_data + result_offset + scanned;
+
+		if (std::is_same<T, PRE_CAST_TYPE>::value) {
+			memcpy(current_result_ptr, scan_state.decompression_buffer + offset_in_compression_group,
+			       to_scan * sizeof(T));
+		} else {
+			for (idx_t i = 0; i < to_scan; i++) {
```

Comment:
Having thought about it a bit more, implementing the bitpacking routines for INT8 and INT16 seems like a good idea. The code size increase should be rather small, since we already implement `1..32` and `1..64`, implementing the additional `1..8` and `1..16` seems like a small overhead. Especially if this is also used for booleans I think this is worth it. Could we do that as well, or would it be a significant extra effort? That way we can unify this code and always decompress straight into the result vector as well.

Reply:
I agree that it would be nice to have this, I'll look into it!